Init Clients Refactor (#33)
datahaven-xyz/datahaven-docs: store-and-retrieve-data/use-storagehub-sdk/upload-a-file.md

## Initialize the Script Entry Point

First, create an `index.ts` file and add the following code:

```ts title="index.ts"
--8<-- 'code/store-and-retrieve-data/use-storagehub-sdk/upload-a-file/upload-a-file.ts:imports'

async function run() {
  // For anything from @storagehub-sdk/core to work, initWasm() is required
  // on top of the file
  await initWasm();
  
  --8<-- 'code/store-and-retrieve-data/use-storagehub-sdk/upload-a-file/upload-a-file.ts:init-setup'

  // --- File upload logic ---
  // **PLACEHOLDER : ADD UPLOAD FILE HELPER METHOD**

  // Disconnect the Polkadot API at the very end
  await polkadotApi.disconnect();
}

await run();
```

## Create Upload File Helper Method

To create the `uploadFile` helper method, first make sure, if you haven't already in the previous guide, to create a new folder called `operations` within the `src` folder (at the same level as the `services` folder) like so:

```bash
mkdir operations
```

Then, create a new file within the `operations` folder called `fileOperations.ts` and add the following code:

```ts title="fileOperations.ts"
--8<-- 'code/store-and-retrieve-data/use-storagehub-sdk/upload-a-file/fileOperations.ts:imports'

> export async function uploadFile(

Reviewer: ```suggestion

export async function uploadFile(
```

In this section of the guide, you’ll go from a local file to a confirmed on-chain transaction. You'll initialize a File Manager, derive the file’s fingerprint, fetch MSP details (including peer IDs), choose a replication level, and issue the storage request. When the transaction is finalized, you’ll have a transaction hash and an on-chain record of the request you can verify in the next section of this guide.

### Initialize File Manager

To initialize the File Manager, add the following code to your `fileOperations.ts` file:

```ts title="fileOperations.ts // **PLACEHOLDER FOR STEP 1: INITIALIZE FILE MANAGER**"
--8<-- 'code/store-and-retrieve-data/use-storagehub-sdk/upload-a-file/fileOperations.ts:initialize-file-manager'
```

### Define Storage Request Parameters

To issue a storage request, you need to prepare the following:
	•	`fingerprint` of your file (from `FileManager`)
	•	`fileSize` in `BigInt` format
	•	`mspId` of the target MSP
	•	`peerId` extracted from the MSP’s multiaddresses
	•	`replicationLevel` that defines how redundancy is applied
	•	`replicas` indicating how many copies to request
	•	`bucketId` created earlier

> 	•	`fileName` you plan to store

Reviewer: This isn't rendering correctly

```

### Retrieve Storage Request Data

To retrieve storage request data, query `fileSystem.storageRequests` and pass in the computed file key:

```ts title="fileOperations.ts // **PLACEHOLDER FOR STEP 5: RETRIEVE STORAGE REQUEST DATA**"
--8<-- 'code/store-and-retrieve-data/use-storagehub-sdk/upload-a-file/fileOperations.ts:verify-storage-request'
```

### Read Storage Request Data

To read storage request data, it first must be unwrapped as follows:

```ts title="fileOperations.ts // **PLACEHOLDER FOR STEP 6: READ STORAGE REQUEST DATA**"
--8<-- 'code/store-and-retrieve-data/use-storagehub-sdk/upload-a-file/fileOperations.ts:read-storage-request'
```

Upon successful storage request verification, you'll see a message like:

--8<-- 'code/store-and-retrieve-data/use-storagehub-sdk/upload-a-file/output-02.html'

??? code "View complete `fileOperations.ts` up until now"

    ```ts title="fileOperations.ts"
    --8<-- 'code/store-and-retrieve-data/use-storagehub-sdk/upload-a-file/fileOperations.ts:imports'
    export async function uploadFile(
    bucketId: string,
    filePath: string,

>     fileName: string

Reviewer: ```suggestion

    export async function uploadFile(
      bucketId: string,
      filePath: string,
      fileName: string
```

---
title: Upload a File

> description: Guide on how to turn your local file into a fully registered and persisted asset in DataHaven.

Reviewer: Meta descriptions should be between 120-160 characters.

1. **Issue a Storage Request**: Register your intent to store a file in your bucket and set its replication policy. Initialize `FileManager`, compute the file’s fingerprint, fetch MSP info (and extract peer IDs), choose a replication level and replica count, then call `issueStorageRequest`.
2. **Verify If Storage Request Is On-Chain**: Derive the deterministic file key, query on-chain state, and confirm the request exists and matches your local fingerprint and bucket.
3. **Upload a File**: Send the file bytes to the MSP, linked to your storage request. Confirm that the upload receipt indicates a successful upload.

These steps form the core workflow for any application that places data into DataHaven.

## Prerequisites

--8<-- 'text/store-and-retrieve-data/use-storagehub-sdk/prerequisites.md'

- [A bucket created](/store-and-retrieve-data/use-storagehub-sdk/create-a-bucket/){target=\_blank} with the ID handy
- A file to upload to DataHaven (any file type is accepted; the current testnet file size limit is {{ networks.testnet.file_size_limit }}).

## Initialize the Script Entry Point

> First, create an `index.ts` file, if you haven't already. Its `run` method will orchestrate all the logic in this guide, and you’ll replace the labelled placeholders with real code step by step. By now, your services folder (including the MSP and client helper services) should already be created. If not, see the [Get Started Guide](/store-and-retrieve-data/use-storagehub-sdk/get-started). The `index.ts` snippet below also imports `fileOperations.ts`, which is not in your project yet. That’s expected, as you’ll create it later in this guide.

Reviewer: ```suggestion
First, create an `index.ts` file if you haven't already. Its `run` method will orchestrate all the logic in this guide, and you’ll replace the labelled placeholder with real code step by step. By now, your services folder (including the MSP and client helper services) should already be created. If not, see the [Get Started](/store-and-retrieve-data/use-storagehub-sdk/get-started/) guide.

The `index.ts` snippet below also imports `fileOperations.ts`, which is not in your project yet—that's expected, as you'll create it later in this guide.
```

```ts title="index.ts"
--8<-- 'code/store-and-retrieve-data/use-storagehub-sdk/upload-a-file/upload-a-file.ts:imports'

async function run() {
  // For anything from @storagehub-sdk/core to work, initWasm() is required
  // on top of the file
  await initWasm();
  
  --8<-- 'code/store-and-retrieve-data/use-storagehub-sdk/upload-a-file/upload-a-file.ts:init-setup'

  // --- File upload logic ---
  // **PLACEHOLDER: ADD UPLOAD FILE HELPER METHOD**

  // Disconnect the Polkadot API at the very end
  await polkadotApi.disconnect();
}

await run();
```

## Add Method to Upload File

Because of the `uploadFile` method's complexity, you will be adding pieces of its logic step by step. Before that, you need to prepare the file and the method's imports, by following these steps:

1. Create a new folder called `operations` within the `src` folder (at the same level as the `services` folder) like so:

    ```bash
    mkdir operations
    ```

2. Create a new file within the `operations` folder called `fileOperations.ts`

3. Add the following code:

    ```ts title="fileOperations.ts"
    --8<-- 'code/store-and-retrieve-data/use-storagehub-sdk/upload-a-file/fileOperations.ts:imports'

>     export async function uploadFile(

Reviewer: Indentation is incorrect here, please fix.

    bucketId: string,
    filePath: string,
    fileName: string
    ) {
    // ISSUE STORAGE REQUEST
    // **PLACEHOLDER FOR STEP 1: INITIALIZE FILE MANAGER**
    // **PLACEHOLDER FOR STEP 2: CREATE FINGERPRINT**
    // **PLACEHOLDER FOR STEP 3: ISSUE STORAGE REQUEST**
    // VERIFY STORAGE REQUEST ON-CHAIN
    // **PLACEHOLDER FOR STEP 4: COMPUTE FILE KEY**
    // **PLACEHOLDER FOR STEP 5: RETRIEVE STORAGE REQUEST DATA**
    // **PLACEHOLDER FOR STEP 6: READ STORAGE REQUEST DATA**
    // UPLOAD FILE
    // **PLACEHOLDER FOR STEP 7: AUTHENTICATE**
    // **PLACEHOLDER FOR STEP 8: UPLOAD FILE TO MSP**

    return { fileKey, uploadReceipt };
    }
    ```

## Issue a Storage Request

A storage request is the instruction that tells DataHaven—through your chosen Main Storage Provider (MSP)—to persist a specific file in a bucket with the redundancy policy you select.

> In this section of the guide, you’ll go from a local file to a confirmed on-chain transaction: initialize a File Manager, derive the file’s fingerprint, fetch MSP details (including peer IDs), choose a replication level, and issue the storage request. When the transaction is finalized, you’ll have a transaction hash and an on-chain record of the request you can verify in the next section of this guide.

Reviewer: ```suggestion
In this section of the guide, you’ll go from a local file to a confirmed on-chain transaction. You’ll initialize a File Manager, derive the file’s fingerprint, fetch MSP details (including peer IDs), choose a replication level, and issue the storage request. When the transaction is finalized, you’ll have a transaction hash and an on-chain record of the request you can verify in the next section of this guide.
```

### Initialize File Manager

To initialize the File Manager, add the following code to your `fileOperations.ts` file:

```ts title="fileOperations.ts // **PLACEHOLDER FOR STEP 1: INITIALIZE FILE MANAGER**"
--8<-- 'code/store-and-retrieve-data/use-storagehub-sdk/upload-a-file/fileOperations.ts:initialize-file-manager'
```

### Define Storage Request Parameters

> To issue a storage request, you must create the `fingerprint` of your file from the File Manager, `fileSize` (in `BigInt` format), `mspId`, `peerId` (extracted from `multiaddresses`), `replicationLevel`, `replicas`, as well as have the `bucketId` and `fileName` handy from before. Add the following code to create all of these:

Reviewer: The first sentence is dense, how can we make it easier to read?
